File: app/src/main/kotlin/com/sponic/langbang/domain/NowVoicing.kt
package com.sponic.langbang.domain

import com.sponic.langbang.data.model.TokenPair
import kotlinx.coroutines.flow.MutableStateFlow

/**
 * What is currently being voiced. Surfaces in a persistent panel so the user can always
 * see the English / Polish / word-for-word literal of whatever audio is playing.
 *
 * `lang` is one of: "en", "pl", "pl-slow", "pause", or null when silent.
 * `words` is the per-token (pl, en) mapping for word-aligned gloss rendering; when
 * absent the panel falls back to whitespace-zipping `literal`.
 */
data class NowVoicing(
    val en: String,
    val pl: String,
    val literal: String?,
    val lang: String,
    val position: String? = null,
    val words: List<TokenPair>? = null
)

object NowVoicingBus {
    val state = MutableStateFlow<NowVoicing?>(null)
    fun publish(v: NowVoicing?) { state.value = v }
    fun clear() { state.value = null }
}

File: web/src/player/nowVoicing.ts
import type { TokenPair } from "../cloud/models";
import { Store } from "../lib/store";

/**
 * What is currently being voiced — surfaced in a persistent panel so the learner can
 * always see the source cue / target answer / word-for-word gloss of the live audio.
 * Web port of the Android NowVoicingBus.
 *
 * `lang`: "en" (source cue), "pl" (target answer), "pl-slow" (slow target), or null when idle.
 */
export interface NowVoicing {
  en: string;
  pl: string;
  literal?: string | null;
  lang: "en" | "pl" | "pl-slow" | "pause" | null;
  position?: string | null;
  words?: TokenPair[] | null;
  /** Hide the target text (recall drill) until reveal. */
  plHidden?: boolean;
}

export const nowVoicing = new Store<NowVoicing | null>(null);

export function clearNowVoicing(): void {
  nowVoicing.set(null);
}
